oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/JwtAuthenticationManagerResolverTests.java

```diff
@@ -0,0 +1,186 @@
+
+			AuthenticationManager authenticationManager =
+					authenticationManagerResolver.resolve(request);
+			assertThat(authenticationManager).isNotNull();
+
+			AuthenticationManager cachedAuthenticationManager =
+					authenticationManagerResolver.resolve(request);
+			assertThat(authenticationManager).isSameAs(cachedAuthenticationManager);
+		}
+	}
+
+	@Test
+	public void resolveWhenUsingUntrustedIssuerThenException() {
+		JwtAuthenticationManagerResolver authenticationManagerResolver =
+				new JwtAuthenticationManagerResolver("other", "issuers");
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Bearer " + this.jwt);
+
+		assertThatCode(() -> authenticationManagerResolver.resolve(request))
+				.isInstanceOf(OAuth2AuthenticationException.class)
+				.hasMessageContaining("Invalid issuer");
+	}
+
+	@Test
+	public void resolveWhenUsingCustomIssuerAuthenticationManagerResolverThenUses() {
+		AuthenticationManager authenticationManager = mock(AuthenticationManager.class);
+		JwtAuthenticationManagerResolver authenticationManagerResolver =
+				new JwtAuthenticationManagerResolver(issuer -> authenticationManager);
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Bearer " + this.jwt);
+
+		assertThat(authenticationManagerResolver.resolve(request))
+				.isSameAs(authenticationManager);
+	}
+
+	@Test
+	public void resolveWhenUsingExternalSourceThenRespondsToChanges() {
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Bearer " + this.jwt);
+
+		Map<String, AuthenticationManager> authenticationManagers = new HashMap<>();
+		JwtAuthenticationManagerResolver authenticationManagerResolver =
+				new JwtAuthenticationManagerResolver(authenticationManagers::get);
+		assertThatCode(() -> authenticationManagerResolver.resolve(request))
+				.isInstanceOf(OAuth2AuthenticationException.class)
+				.hasMessageContaining("Invalid issuer");
+
+		AuthenticationManager authenticationManager = mock(AuthenticationManager.class);
+		authenticationManagers.put("trusted", authenticationManager);
+		assertThat(authenticationManagerResolver.resolve(request))
+				.isSameAs(authenticationManager);
+
+		authenticationManagers.clear();
+		assertThatCode(() -> authenticationManagerResolver.resolve(request))
+				.isInstanceOf(OAuth2AuthenticationException.class)
+				.hasMessageContaining("Invalid issuer");
+	}
+
+	@Test
+	public void resolveWhenBearerTokenMalformedThenException() {
+		JwtAuthenticationManagerResolver authenticationManagerResolver =
+				new JwtAuthenticationManagerResolver("trusted");
+		MockHttpServletRequest request = new MockHttpServletRequest();
+		request.addHeader("Authorization", "Bearer jwt");
+		assertThatCode(() -> authenticationManagerResolver.resolve(request))
+				.isInstanceOf(OAuth2AuthenticationException.class)
+				.hasMessageNotContaining("Invalid issuer");
```

Comment:
Maybe test for the actual exception you're expecting, rather than the one you're _not_ expecting. Or be satisfied that it fails.

Reply:
The message that is returned is one that I don't have control over. I want to confirm that the issuer is not tested itself.